visma/matrix/structure.py

```diff
@@ -103,22 +131,42 @@ def determinant(self):
         pass
 
 
-class IdenMat(SquareMat):
-    """Class for identity matrix
+class DiagMat(SquareMat):
+    """Class for Diagonal matrix
 
-    Identity matrix is a square matrix with all elements as 0 except for the diagonal elements which are 1.
+    Diagonal matrix is a square matrix with all elements as 0 except for the diagonal elements.
 
     Extends:
         SquareMat
     """
 
-    def __init__(self, dim):
+    def __init__(self, dim, diagElem):
+        """
+        dim {list} -- dimension of matrix
+        diagElem {list} -- list of tokens list
+        """
         super().__init__()
+        self.dim = dim
         for i in range(0, dim[0]):
             row = []
             for j in range(0, dim[1]):
                 if i == j:
-                    row.append(Constant(1))
+                    row.append(diagElem[i])
                 else:
-                    row.append(Constant(0))
+                    row.append([Constant(0)])
             self.value.append(row)
+
+
+class IdenMat(DiagMat):
+    """Class for identity matrix
+
+    Identity matrix is a diagonal matrix with all elements as 0 except for the diagonal elements which are 1.
+
+    Extends:
+        DiagMat
+    """
+
+    def __init__(self, dim):
+        super().__init__(dim, ["1"]*dim[0])
```

Comment:
Remove lines 171-172. Idenmat is created from 171 itself.
```suggestion
        super().__init__(dim, [[Constant(1)]]*dim[0])
```